### PR TITLE
add template output

### PR DESCRIPTION
**Pull request by jonasagx**
Allow syft to output an SBOM based on a given [Go template](https://pkg.go.dev/text/template) like Grype currently does. I added a new output format called `template` that receives as param the template file path. I extended the `sbom.Format` to accept format options that formats can use if they implement it the interface themselves. 

Closes: #152 

Signed-off-by: Jonas Xavier <jonasx@anchore.com>

### CODE REVIEW

File: cmd/syft/cli/options/writer.go [hunk 1]
```diff
@@ -55,6 +56,7 @@ func parseOutputs(outputs []string, defaultFile string) (out []sbom.WriterOption
 			errs = multierror.Append(errs, fmt.Errorf("bad output format: '%s'", name))
 			continue
 		}
+		format = format.WithOptions(options.Format{TemplateFilePath: templateFilePath})
```
Comment (wagoodman) [inline cmd/syft/cli/options/writer.go:59]: another way to achieve this without changing the Format interface is to type assert to `template.Format`, and if the assertion is ok, then add on the additional information or create a new instance of the format with the new information injected via a constructor.

File: syft/format-options/options.go [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package options
+
+// Format has options to be used by SBOM formats
+type Format struct {
```
Comment (wagoodman) [inline syft/format-options/options.go:4]: I see why this is here, however, it is a little awkward (it's only purpose is for the state of template and isn't clear how to use this for more formats in the future). Is there a better way to do this?

**Review by wagoodman — COMMENTED** [review]
I left a couple of comments around the` format-options` and downstream effects. The ultimate problem with the code (on `main`) is that there isn't a path to inject dependencies, which is the thing this PR is contending with.

I think there is a compromise that can be made without needing to change the `Format` interface, which is a pattern that affects all formats. That compromise can be made around https://github.com/anchore/syft/pull/1051/files#r900169470 . In the future we can pay down the improvement of injecting all formats from the application config (which solves this whole problem a lot better, but is out of scope for this PR).

File: cmd/syft/cli/options/writer.go (hunk 1 above)
Comment (jonasagx) [reply cmd/syft/cli/options/writer.go]: yes, I like that idea. I used and that helped drop `options.Format`.

File: syft/sbom/format.go [hunk 3]
```diff
@@ -23,6 +25,7 @@ type Format interface {
 	Encode(io.Writer, SBOM) error
 	Decode(io.Reader) (*SBOM, error)
 	Validate(io.Reader) error
+	WithOptions(options.Format) Format
```
[the remark this answers is not in the document]
Comment (jonasagx) [reply syft/sbom/format.go]: I see, I dropped it in favor of type casting to template format.

File: syft/format-options/options.go (hunk 2 above)
Comment (jonasagx) [reply syft/format-options/options.go]: Right, I removed it. Now the template output receives the param directly.

File: README.md [hunk 4]
```diff
@@ -230,6 +230,40 @@ Where the `formats` available are:
 - `spdx-json`: A JSON report conforming to the [SPDX 2.2 JSON Schema](https://github.com/spdx/spdx-spec/blob/v2.2/schemas/spdx-schema.json).
 - `github`: A JSON report conforming to GitHub's dependency snapshot format.
 - `table`: A columnar summary (default).
+- `template`: Lets the user specify the output format. See "Using templates" below.
```
[the remark this answers is not in the document]
Comment (jonasagx) [reply README.md]: I added the link 👍🏼

File: internal/formats/template/format.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	doc := syftjson.ToFormatModel(s)
+	return tmpl.Execute(output, doc)
+}
+
+func (f OutputFormat) Validate(reader io.Reader) error {
+	return sbom.ErrValidationNotSupported
+}
+
+func (f OutputFormat) WithTemplate(filePath string) sbom.Format {
+	return OutputFormat{templateFilePath: filePath}
```
Comment (wagoodman) [inline internal/formats/template/format.go:45]: it's a little odd that this is worded like a functional option, but creates a new object like a builder pattern.

I think this fits a little better as a setter for a field on the instance it was called on, something like
```
f.SetTemplatePath(path)
``` 
(side steps the need to create a new instance of the format).

Comment (jonasagx) [reply internal/formats/template/format.go]: I see now. I've changed the func signature to that.

File: cmd/syft/cli/options/writer.go [hunk 6]
```diff
@@ -55,6 +56,11 @@ func parseOutputs(outputs []string, defaultFile string) (out []sbom.WriterOption
 			errs = multierror.Append(errs, fmt.Errorf("bad output format: '%s'", name))
 			continue
 		}
+		tmpl, ok := format.(template.OutputFormat)
+		if ok {
```
Comment (luhring) [inline cmd/syft/cli/options/writer.go:60]: nit: This is often done on the same line in Go

**Review by luhring — APPROVED** [review]
Looks good!

**Pull request merged**